internal/test/goleak/goleak_test.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (C) MongoDB, Inc. 2024-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.uber.org/goleak"
+)
+
+var dbName = fmt.Sprintf("goleak-%d", time.Now().Unix())
+
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(m)
+}
+
+// TestGoroutineLeak creates clients with various client configurations, runs
+// some operations with each one, then disconnects the client. It asserts that
+// no goroutines were leaked after the client is disconnected.
+func TestGoroutineLeak(t *testing.T) {
+	uri := "mongodb://localhost:27017"
+	if u := os.Getenv("MONGODB_URI"); u != "" {
+		uri = u
+	}
+
+	testCases := []struct {
+		desc string
+		opts *options.ClientOptions
+	}{
+		{
+			desc: "base",
+			opts: options.Client().ApplyURI(uri),
+		},
+		{
+			desc: "compressors=snappy",
+			opts: options.Client().ApplyURI(uri).SetCompressors([]string{"snappy"}),
+		},
+		{
+			desc: "compressors=zlib",
+			opts: options.Client().ApplyURI(uri).SetCompressors([]string{"zlib"}),
+		},
+		{
+			desc: "compressors=zstd",
+			opts: options.Client().ApplyURI(uri).SetCompressors([]string{"zstd"}),
+		},
+		{
+			desc: "minPoolSize=10",
+			opts: options.Client().ApplyURI(uri).SetMinPoolSize(10),
+		},
+		{
+			desc: "serverMonitoringMode=poll",
+			opts: options.Client().ApplyURI(uri).SetServerMonitoringMode(options.ServerMonitoringModePoll),
+		},
+	}
+	for _, tc := range testCases {
+		// These can't be run in parallel because goleak currently can't filter
+		// out goroutines from other parallel subtests.
+		t.Run(tc.desc, func(t *testing.T) {
+			defer goleak.VerifyNone(t)
```

Comment:
The uber-go/goleak documentation implies that we don't have to make this check both in the subtest and at the `TestMain()` level: https://github.com/uber-go/goleak#quick-start . If that's true, suggest removing `TestMain()`.

Reply:
Good idea. I added it to catch goroutine leaks that I didn't know I was looking for, but it's never caught a goroutine leak that `goleak.VerifyNone` didn't catch, so I'm fine removing it.

We should definitely prioritize using `goleak.VerifyNone` in specific test cases because it both indicates which scenario(s) caused a goroutine leak _and_ the output shows up clearly in the Evergreen test logs. `goleak.VerifyTestMain` does neither.

Reply:
Removed `TestMain`.